ambry-router/src/main/java/com/github/ambry/router/TtlUpdateManager.java

```diff
@@ -81,42 +83,63 @@ class TtlUpdateManager {
+      routerMetrics.ttlUpdateBlobNotOriginateLocalOperationRate.mark();
+    }
+
+    // ChunkId list should not contain the metadata blobId itself.
+    while (chunkIdStrs!= null && chunkIdStrs.contains(blobIdStr)) {
+      chunkIdStrs.remove(blobIdStr);
     }
-    if (blobIds.size() == 1) {
+
+    if(chunkIdStrs == null || chunkIdStrs.isEmpty()) {
+      // If there are no chunkIds, then its a simple blob, and we have to update only blob's ttl.
       TtlUpdateOperation ttlUpdateOperation =
-          new TtlUpdateOperation(clusterMap, routerConfig, routerMetrics, blobIds.get(0), serviceId, expiresAtMs,
+          new TtlUpdateOperation(clusterMap, routerConfig, routerMetrics, blobId, serviceId, expiresAtMs,
               time.milliseconds(), callback, time, futureResult, quotaChargeCallback);
       ttlUpdateOperations.add(ttlUpdateOperation);
-    } else {
-      BatchOperationCallbackTracker tracker =
-          new BatchOperationCallbackTracker(blobIds, futureResult, callback, quotaChargeCallback);
-      long operationTimeMs = time.milliseconds();
-      for (BlobId blobId : blobIds) {
-        TtlUpdateOperation ttlUpdateOperation =
-            new TtlUpdateOperation(clusterMap, routerConfig, routerMetrics, blobId, serviceId, expiresAtMs,
-                operationTimeMs, tracker.getCallback(blobId), time, BatchOperationCallbackTracker.DUMMY_FUTURE,
-                quotaChargeCallback);
-        ttlUpdateOperations.add(ttlUpdateOperation);
+      return;
+    }
+
+    // If we are here, that means the blob is a composite blob. So we will do a batch operation.
+    List<BlobId> chunkIds = new ArrayList<>();
+    for (String chunkIdStr : chunkIdStrs) {
+      BlobId chunkId = RouterUtils.getBlobIdFromString(chunkIdStr, clusterMap);
+      if (chunkId.getDatacenterId() != ClusterMap.UNKNOWN_DATACENTER_ID
+          && chunkId.getDatacenterId() != clusterMap.getLocalDatacenterId()) {
+        routerMetrics.ttlUpdateBlobNotOriginateLocalOperationRate.mark();
       }
+      chunkIds.add(chunkId);
+    }
+
+    BatchOperationCallbackTracker tracker =
+        new BatchOperationCallbackTracker(chunkIds, blobId, futureResult, callback, quotaChargeCallback, (b) -> {
+          TtlUpdateOperation ttlUpdateOperation =
+              new TtlUpdateOperation(clusterMap, routerConfig, routerMetrics, b, serviceId, expiresAtMs,
+                  time.milliseconds(), callback, time, futureResult, quotaChargeCallback);
+          ttlUpdateOperations.add(ttlUpdateOperation);
+        });
+    long operationTimeMs = time.milliseconds();
+    for (BlobId chunkId : chunkIds) {
+      TtlUpdateOperation ttlUpdateOperation =
+          new TtlUpdateOperation(clusterMap, routerConfig, routerMetrics, chunkId, serviceId, expiresAtMs,
+              operationTimeMs, tracker.getCallback(chunkId), time, BatchOperationCallbackTracker.DUMMY_FUTURE,
+              quotaChargeCallback);
+      ttlUpdateOperations.add(ttlUpdateOperation);
```

Comment:
I think we should move the final blob id to the callback of the chunk ids. something like

```
FutureResult chunkFutureResult = new FutureResult<Void>();
BatchOperationCallbackTracker tracker = new BatchOperationCallbackTracker(chunkIds, chunkFutureResult, (r, e) -> {
    TtlUpdateOperation operationForFinalBlob = new TtlUpdateOperation(clusterMap, routerConfig, routerMetrics, blobId, serviceId, expiresAtMs, time.milliseconds(), callback, time, futureResult, quotaChargeCallback);
    ttlUpdateOperations.add(operationForFinalBlob);
}, quotaChargeCallback);

for (BlobId chunkId: chunkIds) {
    TtlUpdateOperation ttlUpdateOperation =
          new TtlUpdateOperation(clusterMap, routerConfig, routerMetrics, chunkId, serviceId, expiresAtMs,
              operationTimeMs, tracker.getCallback(chunkId), time, BatchOperationCallbackTracker.DUMMY_FUTURE,
              quotaChargeCallback);
}
```

And then we don't have to change anything in BatchOperationCallbackTracker.

Reply:
I think it won't be this simple. The callback of BatchOperationTracker also needs to take care of any exception thrown. So we have to handle the `(e == null)` case as well. Its cleaner (for future enchancement purposes too), to keep all the tracking logic in the `BatchOperationCallbackTracker`.

Reply:
Still think using the callback to handle the metadata blob id is the right way to do it. But if you think otherwise, that's fine.